src/QueryM/WaitUntilSlot.purs

```diff
@@ -137,3 +142,39 @@ posixTimeToSeconds (POSIXTime futureTimeBigInt) = do
     $ map (wrap <<< Int.toNumber)
     $ BigInt.toInt
     $ futureTimeBigInt / BigInt.fromInt 1000
+
+-- | Wait at least `offset` number of slots.
+waitNSlots :: Natural -> QueryM Chain.Tip
+waitNSlots offset = do
+  offsetBigNum <- liftM (error "Unable to convert BigInt to BigNum")
+    $ (BigNum.fromBigInt <<< Natural.toBigInt) offset
+  if offsetBigNum == BigNum.fromInt 0 then getChainTip
+  else do
+    slot <- currentSlot
+    newSlot <- liftM (error "Unable to advance slot")
+      $ wrap <$> BigNum.add (unwrap slot) offsetBigNum
+    waitUntilSlot newSlot
+
+currentSlot :: QueryM Slot
+currentSlot = getChainTip >>= case _ of
+  Chain.Tip (Chain.ChainTip { slot }) -> pure slot
+  Chain.TipAtGenesis -> (pure <<< Slot <<< BigNum.fromInt) 0
+
+-- | Get the latest POSIXTime of the current slot.
+-- The plutus implementation relies on `slotToEndPOSIXTime`
+-- https://github.com/input-output-hk/plutus-apps/blob/fb8a39645e532841b6e38d42ecb957f1945833a5/plutus-contract/src/Plutus/Contract/Trace.hs
+currentTime :: QueryM POSIXTime
+currentTime = currentSlot >>= slotToEndPOSIXTime
+
+-- | Get the ending 'POSIXTime' of a 'Slot' related to
+-- | our `QueryM` configuration.
+-- see https://github.com/input-output-hk/plutus-apps/blob/fb8a39645e532841b6e38d42ecb957f1945833a5/plutus-ledger/src/Ledger/TimeSlot.hs
+slotToEndPOSIXTime :: Slot -> QueryM POSIXTime
+slotToEndPOSIXTime slot = do
+  futureSlot <- liftM (error "Unable to advance slot")
+    $ wrap <$> BigNum.add (unwrap slot) (BigNum.fromInt 1)
+  eraSummaries <- getEraSummaries
+  sysStart <- getSystemStart
+  futureTime <- liftEffect $ slotToPosixTime eraSummaries sysStart futureSlot
+    >>= hush >>> liftM (error "Unable to convert Slot to POSIXTime")
+  pure ((+) (wrap <<< BigInt.fromInt $ -1) futureTime)
```

Comment:
Not super important but I'm curious if there's a reason to choose prefix `+` instead of infix?

Reply:
I think at some point I  was trying to `(+ a) <$>`  ala Haskell but then I discovered it has to be `(+) a <$>` and then I ended removing the `<$>`.